deis/controller
# feat(app): during ps:restart run pod delete in parallel

Uses the asyncio module to execute sync functions in parallel

Closes #404

Gives a bit of a speed boost when restarting 20 pods:

Without asyncio

```
$ time deis ps:restart cmd -a earthy-underdog
Restarting processes... but first, coffee!
done in 39s
=== earthy-underdog Processes
--- cmd:
earthy-underdog-v2-cmd-uby76 up (v2)
earthy-underdog-v2-cmd-c9s0c up (v2)
earthy-underdog-v2-cmd-s89aa up (v2)
earthy-underdog-v2-cmd-mbusz up (v2)
earthy-underdog-v2-cmd-cjs75 up (v2)
earthy-underdog-v2-cmd-k7pit up (v2)
earthy-underdog-v2-cmd-4gdsb up (v2)
earthy-underdog-v2-cmd-r0g9k up (v2)
earthy-underdog-v2-cmd-9ep20 up (v2)
earthy-underdog-v2-cmd-ygmo0 up (v2)
earthy-underdog-v2-cmd-orhgn up (v2)
earthy-underdog-v2-cmd-7t2rb up (v2)
earthy-underdog-v2-cmd-vkxrx up (v2)
earthy-underdog-v2-cmd-mz6nt up (v2)
earthy-underdog-v2-cmd-gg74e up (v2)
earthy-underdog-v2-cmd-1a1hu up (v2)
earthy-underdog-v2-cmd-vmrmp up (v2)
earthy-underdog-v2-cmd-gsiv6 up (v2)
earthy-underdog-v2-cmd-aq800 up (v2)
earthy-underdog-v2-cmd-74lrv up (v2)

real	0m39.545s
user	0m0.013s
sys	0m0.019s
```

With asyncio

```
time deis ps:restart cmd -a earthy-underdog
Restarting processes... but first, coffee!
done in 22s
=== earthy-underdog Processes
--- cmd:
earthy-underdog-v2-cmd-7y1gv up (v2)
earthy-underdog-v2-cmd-qufj4 up (v2)
earthy-underdog-v2-cmd-jxhae up (v2)
earthy-underdog-v2-cmd-q9eh2 up (v2)
earthy-underdog-v2-cmd-1fs34 up (v2)
earthy-underdog-v2-cmd-au0kq up (v2)
earthy-underdog-v2-cmd-k4z4r up (v2)
earthy-underdog-v2-cmd-21xc8 up (v2)
earthy-underdog-v2-cmd-siafb up (v2)
earthy-underdog-v2-cmd-romjl up (v2)
earthy-underdog-v2-cmd-ewhyp up (v2)
earthy-underdog-v2-cmd-va1qs up (v2)
earthy-underdog-v2-cmd-jtfp7 up (v2)
earthy-underdog-v2-cmd-dduk8 up (v2)
earthy-underdog-v2-cmd-w28lt up (v2)
earthy-underdog-v2-cmd-9ovrj up (v2)
earthy-underdog-v2-cmd-ajvh4 up (v2)
earthy-underdog-v2-cmd-j7sxl up (v2)
earthy-underdog-v2-cmd-mpobo up (v2)
earthy-underdog-v2-cmd-u4xy2 up (v2)

real	0m22.996s
user	0m0.011s
sys	0m0.019s
```

Not always this fast as the whole command also accounts for waiting for new pods to come into existance. Single pod restart performance is not improved with this.

Note that I kept some information around what pods are being deleted and if they are deleted in debug mode, so best way to test this is turning `DEIS_DEBUG` on in the controller RC

The test plan roughly goes:

* Create an app and deploy with `deis/example-go`
* scale to 20 or 10, a big enough of a number
* `deis ps:restart -a <app>`
* Look at logs with debug on
* observe the speed increase between this branch and master
